state-chain/chains/src/refund_parameters.rs

```diff
@@ -0,0 +1,255 @@
+pub type ChannelRefundParametersUncheckedInternalAddress =
+	ChannelRefundParametersUnchecked<ForeignChainAddress>;
+
+/// The legacy refund parameters type that does not include any ccm refund metadata.
+pub type ChannelRefundParametersLegacy<RefundAddress> = ChannelRefundParameters<RefundAddress, ()>;
+
+/// Convenience alias for unchecked refund parameters with encoded addresses. This is most commonly
+/// used in State Chain events and extrinsics.
+pub type ChannelRefundParametersUncheckedEncoded = ChannelRefundParametersUnchecked<EncodedAddress>;
+
+/// This is the type used in internal APIs, for example when making swap request.
+pub type ChannelRefundParametersCheckedInternal<AccountId> =
+	ChannelRefundParametersChecked<AccountOrAddress<AccountId, ForeignChainAddress>>;
+
+/// Convenience alias for unchecked refund parameters with the refund address as an `AddressString`.
+/// This is used in RPCs where we require the refund address to be (de)serializable.
+#[cfg(feature = "std")]
+pub type RpcChannelRefundParameters =
+	ChannelRefundParametersUnchecked<crate::address::AddressString>;
+
+/// Convenience alias for unchecked refund parameters with the refund address as the chain's account
+/// type.
+pub type ChannelRefundParametersForChain<C> =
+	ChannelRefundParametersUnchecked<<C as Chain>::ChainAccount>;
+
+impl<A, D> ChannelRefundParameters<A, D> {
+	pub fn min_output_amount(&self, input_amount: AssetAmount) -> AssetAmount {
+		use sp_runtime::traits::UniqueSaturatedInto;
+		cf_amm_math::output_amount_ceil(input_amount.into(), self.min_price).unique_saturated_into()
+	}
+}
+
+#[cfg(feature = "std")]
+impl RpcChannelRefundParameters {
+	pub fn parse_refund_address_for_chain(
+		self,
+		chain: cf_primitives::ForeignChain,
+	) -> anyhow::Result<ChannelRefundParametersUncheckedEncoded> {
+		Ok(ChannelRefundParameters {
+			retry_duration: self.retry_duration,
+			refund_address: self.refund_address.try_parse_to_encoded_address(chain)?,
+			min_price: self.min_price,
+			refund_ccm_metadata: self.refund_ccm_metadata.clone(),
+		})
+	}
+}
+
+impl<A, D> ChannelRefundParameters<A, D> {
+	pub fn map_address<B, F: FnOnce(A) -> B>(self, f: F) -> ChannelRefundParameters<B, D> {
+		ChannelRefundParameters {
+			retry_duration: self.retry_duration,
+			refund_address: f(self.refund_address),
+			min_price: self.min_price,
+			refund_ccm_metadata: self.refund_ccm_metadata,
+		}
+	}
+	pub fn try_map_address<B, E, F: FnOnce(A) -> Result<B, E>>(
+		self,
+		f: F,
+	) -> Result<ChannelRefundParameters<B, D>, E> {
+		Ok(ChannelRefundParameters {
+			retry_duration: self.retry_duration,
+			refund_address: f(self.refund_address)?,
+			min_price: self.min_price,
+			refund_ccm_metadata: self.refund_ccm_metadata,
+		})
+	}
+}
+
+impl<A> ChannelRefundParametersUnchecked<A> {
+	/// Converts the refund address into the ForeignChainAddress type.
+	pub fn with_foreign_chain_refund_address<C>(
```

Comment:
How about:
```suggestion
	pub fn decode_chain_refund_address<C>(
```

Otherwise "with_" makes it sound like we adding something to the object rather than converting an existing field.